astroquery/tests/test_cache.py

```diff
@@ -0,0 +1,222 @@
+
+    with conf.set_temp("cache_timeout", 5):
+        assert conf.cache_timeout == 5
+
+    with conf.set_temp("cache_active", False):
+        assert conf.cache_active is False
+
+    assert conf.cache_timeout == default_timeout
+    assert conf.cache_active == default_active
+
+    conf.cache_timeout = 5
+    conf.cache_active = False
+    conf.reset()
+
+    assert conf.cache_timeout == default_timeout
+    assert conf.cache_active == default_active
+
+
+def test_basic_caching():
+    conf.reset()
+
+    mytest = TestClass()
+    assert conf.cache_active
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    set_response(TEXT1)
+
+    resp = mytest.test_func(URL1)
+    assert resp.content == TEXT1
+    assert len(os.listdir(mytest.cache_location)) == 1
+
+    set_response(TEXT2)
+
+    resp = mytest.test_func(URL2)  # query that has not been cached
+    assert resp.content == TEXT2
+    assert len(os.listdir(mytest.cache_location)) == 2
+
+    resp = mytest.test_func(URL1)
+    assert resp.content == TEXT1  # query that was cached
+    assert len(os.listdir(mytest.cache_location)) == 2  # no new cache file
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    resp = mytest.test_func(URL1)
+    assert resp.content == TEXT2  # Now get new response
+
+
+def test_change_location(tmpdir):
+    conf.reset()
+
+    mytest = TestClass()
+    default_cache_location = mytest.cache_location
+
+    assert paths.get_cache_dir() in str(default_cache_location)
+    assert "astroquery" in mytest.cache_location.parts
+    assert mytest.name in mytest.cache_location.parts
+
+    new_loc = "new_dir"
+    mytest.cache_location = new_loc
+    assert str(mytest.cache_location) == new_loc
+
+    mytest.reset_cache_location()
+    assert mytest.cache_location == default_cache_location
+
+    Path(new_loc).mkdir(parents=True, exist_ok=True)
+    with paths.set_temp_cache(new_loc):
+        assert new_loc in mytest.cache_location.parts
+        assert "astroquery" in mytest.cache_location.parts
+        assert mytest.name in mytest.cache_location.parts
+
+
+def test_login():
+    conf.reset()
+
+    mytest = TestClass()
+    assert conf.cache_active
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    set_response(TEXT1)  # Text 1 is set as the approved password
+
+    mytest.login("ceb")
+    assert mytest.authenticated()
+    assert len(os.listdir(mytest.cache_location)) == 0  # request should not be cached
+
+    set_response(TEXT2)  # Text 2 is not the approved password
+
+    mytest.login("ceb")
+    assert not mytest.authenticated()  # Should not be accessing cache
+
+
+def test_timeout():
+    conf.reset()
+
+    mytest = TestClass()
+    assert conf.cache_active
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    conf.cache_timeout = 2  # Set to 2 sec so we can reach timeout easily
```

Comment:
Instead of setting the `cache_timeout` to be small and sleeping for long enough it would be better to monkeypatch `datetime.utcfromtimestamp()`.

Reply:
I'm struggling to do this in a way that allows me to turn it on/off at will. @bsipocz  Do you have any suggestions for avoiding the sleep here?

Reply:
Based on [`pytest` documentation](https://docs.pytest.org/en/7.1.x/how-to/monkeypatch.html#monkeypatching-functions) something like this should work:
```python
from datetime import datetime
...
def test_expired_cache(monkeypatch):
    def mockreturn(filename):
        return datetime(1970, 1, 1)

    monkeypatch.setattr(datetime, "utcfromtimestamp", mockreturn)
    ...
```

Reply:
Turns out that doesn't work because python does not let you set attributes of built-ins. Instead of sorting out a way around that, I changed the date on the file itself which worked a treat.

Reply:
Good. Avoiding the `sleep` call was the important thing here.